src/array.cr

```diff
@@ -2029,6 +2029,31 @@ class Array(T)
     @capacity - @offset_to_buffer
   end
 
+  private def rewind
+    root_buffer.copy_from(@buffer, @size)
+    shift_buffer_by -@offset_to_buffer
+  end
+
+  # Enusures that the internal buffer has at least `capacity` elements.
+  def ensure_capacity(capacity : Int32) : self
+    rewind
```

Comment:
This will do a realloc even if the given `capacity` is smaller than `@size`, is that intended?

Reply:
It doesn't realloc, but memcpy. But that's the same problem. 
I'm not sure `#ensure_capacity` should ever rewind.

Reply:
There's never a need to rewind.